saleor/graphql/core/connection.py

```diff
@@ -1,7 +1,118 @@
+from typing import Any, Dict
+
 import graphene
+from django.db.models.query import QuerySet
 from graphene import Field, List, NonNull, ObjectType, String
 from graphene.relay.connection import Connection
 from graphene_django_optimizer.types import OptimizedDjangoObjectType
+from graphql_relay.connection.connectiontypes import Edge, PageInfo
+
+ConnectionArguments = Dict[str, Any]
+
+
+def connection_args_validation(args):
+    first = args.get("first")
+    last = args.get("last")
+
+    if isinstance(first, int) and isinstance(last, int):
+        raise ValueError("Argument 'last' cannot be combined with 'first'.")
+    if isinstance(first, int) and args.get("before"):
+        raise ValueError("Argument 'first' cannot be combined with 'before'.")
+    if isinstance(last, int) and args.get("after"):
+        raise ValueError("Argument 'last' cannot be combined with 'after'.")
+    if isinstance(first, int) and first < 0 or first and not isinstance(first, int):
+        raise ValueError("Argument 'first' must be a non-negative integer.")
+    if isinstance(last, int) and last < 0 or last and not isinstance(last, int):
+        raise ValueError("Argument 'last' must be a non-negative integer.")
+
+
+def connection_from_queryset_slice(
+    qs: QuerySet,
+    args: ConnectionArguments = None,
+    connection_type: Any = Connection,
+    edge_type: Any = Edge,
+    page_info_type: Any = PageInfo,
+) -> Connection:
+    """Create a connection object from a QuerySet."""
+    args = args or {}
+    before = args.get("before")
+    after = args.get("after")
+    first = args.get("first")
+    last = args.get("last")
+
+    connection_args_validation(args)
+
+    cursor_after = graphene.Node.from_global_id(after) if after else None
+    cursor_before = graphene.Node.from_global_id(before) if before else None
+    cursor_model = qs.model
+    cursor = cursor_after or cursor_before
+    cursor_offset = 1 if cursor else 0
+
+    sort_by = args.get("sort_by", {})
+    sort_by_field = sort_by.get("field")
+    sort_by_direction = sort_by.get("direction", "")
```

Comment:
I think these should be an enum and a boolean in the input type. I like Shopify's simple interface:

```graphql
{
  products(first: 50, sortKey: CREATED_AT, reverse: true) {
    ...
  }
}
```

Reply:
Giving it a second thought I think the mutation should take care of mapping from `sortKey` to `sort_field` (available choices will depend on the collection being paginated) so this function could accept `sort_field: str, reverse: bool`.

Reply:
On `master` we have `products(sortBy: {direction: ASC, field: NAME})`. Does this PR change it or you don't want to have it inside the `sortBy` input object @patrys ?

Reply:
Disregard my comment then. It does look like passing any direction would result in the order being reversed though.